File: advent-of-code/2023/day-01/part2.py
#!/opt/homebrew/bin/python3

'''
Your calculation isn't quite right_index. It looks like some of the digits 
are actually spelled out with letters: one, two, three, four, five, 
six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real 
first and last digit on each line. For example:

two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. 
Adding these together produces 281.

approach:
    - first check if letter is a digit
    - then build a trie to see if the cut off is in word set
'''

import argparse

DIGIT_MAP = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9'
}

key_lengths = set([len(k) for k in DIGIT_MAP.keys()])

def get_digit(start_index, line):
    if line[start_index].isdigit():
        return line[start_index]

    
    for k in key_lengths:
        if (
                start_index + k - 1 < len(line) and 
                line[start_index:start_index+k] in DIGIT_MAP
        ):
            return DIGIT_MAP[line[start_index:start_index+k]]

    return None


def solve(lines):
    total = 0
    for l in lines:
        start_digit, end_digit = None, None

        # get start digit
        for left_index in range(len(l)):
            start_digit = get_digit(start_index=left_index, line=l)
            if start_digit:
                break

        if not start_digit:
            raise Exception(f"there was not anything digit in line {l}")

        for right_index in range(len(l) - 1, -1, -1):
            end_digit = get_digit(start_index=right_index, line=l)
            if end_digit:
                break

        total += int(start_digit + end_digit)
    return total


def main():
    parser = argparse.ArgumentParser(description='Advent of code day 1 part 2')
    parser.add_argument('-i', '--input', required=True)
    args = parser.parse_args()

    try:
        with open(args.input, 'r') as input_file:
            lines = [line.strip() for line in input_file]
            print(solve(lines))
    except FileNotFoundError:
        print(f"Input file not found: {args.input}")
    except Exception as e:
        print(f"Encounter error for input: {args.input}, err: {e}")

if __name__ == "__main__":
    main()

File: advent-of-code/2023/day-01/part2.go
package main

import (
  "bufio"
  "flag"
  "fmt"
  "os"
  "errors"
  "strconv"
)

const (
  BAD_CANDIDATE = "BAD_CANDIDATE"
)

func main() {
  var inputFile string
  flag.StringVar(&inputFile, "i", "", "Specify the input filename")
  flag.StringVar(&inputFile, "input", "", "Specify the input filename")

  flag.Parse()

  if inputFile == "" {
    fmt.Println("Usage: go run part1.go -i/--input <filename>")
    return
  }

  file, err := os.Open(inputFile)
  if err != nil {
    fmt.Println("Error opening file:", err)
    return
  }
  defer file.Close()

  var lines []string

  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    line := scanner.Text()
    lines = append(lines, line)
  }

  if err := solve(lines); err != nil {
    fmt.Println("Error trying to solve:", err)
  }
  

  if err := scanner.Err(); err != nil {
    fmt.Println("Error reading file:", err)
    return
  }
}

// check if candidate is good and return the candidate word, then find the corresponding mapping
func getCandidate(line string, startIndex int) string {
  digitWords := map[string]int{
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
  }
  lineLength := len(line)
  for key, _ := range digitWords {
    if startIndex + len(key) - 1 < lineLength {
      digitWordCandidate := line[startIndex:startIndex+len(key)]
      if digitWordCandidate == key {
        return key
      }
    }
  }
  return BAD_CANDIDATE
}

func solve(lines []string) error {
  digitWords := map[string]int{
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
  }
  var total = 0;
  for _, l := range lines {
    var startDigit int
    var endDigit int

    // get first digit
    for i := 0; i < len(l); i++ {
      if '0' <= l[i] && l[i] <= '9' {
        startDigit, _ = strconv.Atoi(l[i:i+1])
        break
      } 
      result := getCandidate(l, i)
      if result != BAD_CANDIDATE {
        startDigit = digitWords[result]
        break
      }
    }
    if startDigit == 0 {
      errorMessage := fmt.Sprintf("cannot find a digit from line: %s", l)
      return errors.New(errorMessage)
    }

    // get second digit
    for i := len(l) - 1; i >= 0; i-- {
      if '0' <= l[i] && l[i] <= '9' {
        endDigit, _ = strconv.Atoi(l[i:i+1])
        break
      }
      result := getCandidate(l, i)
      if result != BAD_CANDIDATE {
        endDigit = digitWords[result]
        break
      }
    }

    total += (startDigit * 10 + endDigit)
  }
  fmt.Println(total)
  return nil
}


